verify/verify.go

```diff
@@ -109,9 +109,19 @@ func (db *DB) VerifySignatures(s *data.Signed, role string) error {
 
 		// Only consider this key valid if we haven't seen any of it's
 		// key ids before.
-		if _, ok := seen[sig.KeyID]; !ok {
-			for _, id := range verifier.MarshalPublicKey().IDs() {
-				seen[id] = struct{}{}
+		// Careful: we must not rely on the key IDs _declared in the file_,
+		// instead we get to decide what key IDs this key correspond to.
+		// XXX dangerous; better stop supporting multiple key IDs altogether.
+		keyIDs := verifier.MarshalPublicKey().IDs()
+		wasKeySeen := false
```

Comment:
Optimisation nit: we should perform this check _before_ we even verify the signature.

Reply:
What is the expected behavior if there are two signatures from the same keyid, but one is invalid? I think leaving this here ensures that the valid signature would be used, whether or not it is the first one in the list

Reply:
You are right: leaving this alone.